src/imitation/util/buffering_wrapper.py

```diff
@@ -0,0 +1,75 @@
+import numpy as np
+from stable_baselines.common.vec_env import VecEnv, VecEnvWrapper
+
+from imitation.util import rollout
+
+
+class BufferingWrapper(VecEnvWrapper):
```

Comment:
It feels like this is doing something very similar to `imitation.util.rollout`. Can we combine them in some way to avoid code duplication?

One approach would be to use `BufferingWrapper` inside `imitation.util.rollout`, to store the trajectories.

Currently there's an implementation difference that `BufferingWrapper` stores transitions and `rollout` works on the basis of trajectories, but we could change `BuffferingWrapper` to be trajectories (just port over code from `rollout.generate_trajectories`).

Reply:
As you noted it could be tricky to use the current `rollout.Transition`-based BufferingWrapper within `generate_trajectories`. I'm wondering right now if it's important that BufferingWrapper is Transition rather than Trajectory-based.

If a typical trajectory has many more timesteps than `buffer.capacity // n_vec` then at the very least this would lead to a `ValueError("Not enough capacity to store data.")`.

It could also be troublesome I think if there's epochs in the beginning of training where the buffer isn't updated.

Reply:
Looking through `_TrajectoryAccumulator` a second time, I think it actually would be a straightforward refactor to allow `BufferingWrapper` to either `pop_trajectories` or `pop_transitions`.

We would use `TrajectoryAccumulator` inside `BufferingWrapper`, and simply do something similar to `finish_trajectory` on every unfinished intermediate trajectory data structure if the user calls `pop_transitions`.

Reply:
The transitions vs trajectory thing is a pain. It seems the main issue is that we want to return trajectory segments that are not complete? Calling `finish_trajectory` prematurely seems fine.